frontend/src/scenes/pipeline/hogFunctionTestingLogic.tsx

```diff
@@ -0,0 +1,333 @@
+                        } satisfies EventsNode,
+                    ],
+                    properties: matchingFilters,
+                    dateRange: {
+                        date_from: dateRange.after,
+                        date_to: dateRange.before,
+                    },
+                    trendsFilter: {
+                        display: ChartDisplayType.BoldNumber,
+                    },
+                }
+            },
+            { resultEqualityCheck: equal },
+        ],
+        eventsWithRetries: [
+            (s) => [s.events, s.retries],
+            (events: { results: any[] }, retries: HogFunctionTestInvocationResultWithEventId[]) =>
+                events.results.map((row) => [
+                    ...row.slice(0, 3),
+                    retries.filter((r) => r.eventId === row[0].uuid),
+                    ...row.slice(3),
+                ]),
+        ],
+        nextQuery: [
+            (s) => [s.baseEventsQuery, s.events],
+            (baseEventsQuery, events): EventsQuery | null => {
+                if (!baseEventsQuery || !events) {
+                    return null
+                }
+                const typedResults = events?.results
+                const sortColumnIndex = baseEventsQuery?.select
+                    .map((hql: any) => removeExpressionComment(hql))
+                    .indexOf('timestamp')
+                if (sortColumnIndex !== -1) {
+                    const lastTimestamp = typedResults?.[typedResults.length - 1]?.[sortColumnIndex]
+                    if (lastTimestamp) {
+                        const newQuery: EventsQuery = {
+                            ...baseEventsQuery,
+                            before: lastTimestamp,
+                            limit: PAGE_ROWS,
+                        }
+                        return newQuery
+                    }
+                }
+                return null
+            },
+        ],
+        previousQuery: [
+            (s) => [s.baseEventsQuery, s.events],
+            (baseEventsQuery, events): EventsQuery | null => {
+                if (!baseEventsQuery || !events) {
+                    return null
+                }
+                const lastTimestamp = values.pageTimestamps[values.pageTimestamps.length - 1]
+                const newQuery: EventsQuery = {
+                    ...baseEventsQuery,
+                    before: lastTimestamp,
+                    limit: PAGE_ROWS,
+                }
+                return newQuery
+            },
+        ],
+    })),
+    listeners(({ actions }) => ({
+        changeDateRange: () => {
+            actions.loadEvents()
+            actions.loadTotalEvents()
+            actions.resetCurrentPage()
+            actions.resetSelectedForRetry()
+        },
+        loadEvents: () => {
+            actions.loadTotalEvents()
+            actions.resetCurrentPage()
+            actions.resetCollapsedRows()
+            actions.resetSelectedForRetry()
+        },
```

Comment:
changeDateRange calls loadEvents and then loadEvents is calling all the same actions?